### PR TITLE
ci: Add GH workflow to build app & deploy to GH Pages; Update demo link to point at GH Pages deployment.

### PR DESCRIPTION
**Pull request by junhaoliao**
# References

Resolves https://github.com/y-scope/yscope-log-viewer/issues/102#issuecomment-2455503426

# Description

1. Add workflow to build the application and deploy to GitHub pages.
2. Update demo link in README.md to refer to the deployment.

# Validation performed

On @junhaoliao 's fork:
1. Observed the GitHub workflow completed without error: https://github.com/junhaoliao/yscope-log-viewer/actions/runs/11752079920
3. Accessed the application at https://junhaoliao.github.io/yscope-log-viewer/ which was successful.

## Summary by CodeRabbit

- **New Features**
  - Updated the online demo URL for the log viewer to ensure users access the correct location.
  - Introduced a new GitHub Actions workflow for automated deployment to GitHub Pages.

- **Documentation**
  - Revised the README.md to reflect the updated online demo link.

### CODE REVIEW

File: .github/workflows/deploy-gh-pages.yml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    - uses: "actions/checkout@v4"
+
+    - name: "Set up Node.js"
+      uses: "actions/setup-node@v4"
+      with:
+        node-version: "22"
+
+    - run: "npm install"
+    - run: "npm run build"
+    - name: "Upload static files as artifact"
```
Comment (junhaoliao) [inline .github/workflows/deploy-gh-pages.yml:21]:
```suggestion

    - name: "Upload static files as artifact"
```

File: .github/workflows/deploy-gh-pages.yml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    permissions:
+      pages: "write"     # to deploy to Pages
+      id-token: "write"  # to verify the deployment originates from an app
+    runs-on: "ubuntu-latest"
+    environment:
+      name: "github-pages"
+      url: ${{ steps.deployment.outputs.page_url }}
+    steps:
+      - name: Deploy to GitHub Pages
+        uses: actions/deploy-pages@v4
```
Comment (junhaoliao) [inline .github/workflows/deploy-gh-pages.yml:38]:
```suggestion
      url: "${{steps.deployment.outputs.page_url}}"
    steps:
      - name: "Deploy to GitHub Pages"
        uses: "actions/deploy-pages@v4"
```

File: .github/workflows/deploy-gh-pages.yml [hunk 3]
```diff
@@ -0,0 +1,42 @@
+name: "Deploy to GitHub Pages"
```
Comment (junhaoliao) [inline .github/workflows/deploy-gh-pages.yml:1]:
```suggestion
name: "deploy-github-pages"
```

File: .github/workflows/deploy-gh-pages.yml [hunk 4]
```diff
@@ -0,0 +1,42 @@
+name: "deploy-github-pages"
+
+on:
+  push:
+    branches: [ "main" ]
```
Comment (kirkrodrigues) [inline .github/workflows/deploy-gh-pages.yml:5]:
```suggestion
    branches: ["main"]
```

File: .github/workflows/deploy-gh-pages.yml [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+jobs:
+  build:
+    runs-on: "ubuntu-latest"
+
```
Comment (kirkrodrigues) [inline .github/workflows/deploy-gh-pages.yml:10]:
```suggestion
```

File: .github/workflows/deploy-gh-pages.yml [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+      - uses: "actions/checkout@v4"
+
+      - uses: "actions/setup-node@v4"
+        with:
+          node-version: "22"
+          cache: "npm"
+          cache-dependency-path: "./package-lock.json"
+
+      - run: "npm install"
+      - run: "npm run build"
```
Comment (kirkrodrigues) [inline .github/workflows/deploy-gh-pages.yml:21]:
```suggestion

      - run: "npm run build"
```

File: .github/workflows/deploy-gh-pages.yml [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+      - name: "Upload static files as artifact"
+        uses: "actions/upload-pages-artifact@v3"
+        with:
+          path: "./dist"
+
+  deploy:
+    needs: "build"
+    permissions:
+      # to deploy to Pages
```
Comment (kirkrodrigues) [inline .github/workflows/deploy-gh-pages.yml:31]:
```suggestion
      # To deploy to Pages
```

File: .github/workflows/deploy-gh-pages.yml [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+      pages: "write"
+      # to verify the deployment originates from an app
```
Comment (kirkrodrigues) [inline .github/workflows/deploy-gh-pages.yml:33]:
```suggestion

      # To request a JWT from GitHub
```

Based on https://docs.github.com/en/actions/security-for-github-actions/security-hardening-your-deployments/about-security-hardening-with-openid-connect#adding-permissions-settings

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

File: .github/workflows/deploy-gh-pages.yml (hunk 8 above)
Comment (junhaoliao) [reply .github/workflows/deploy-gh-pages.yml:33]: How about
```suggestion

      # To request a JWT from GitHub to verify the deployment originates from an appropriate source
```
I was referring to this: https://github.com/actions/deploy-pages#:~:text=deploy%20to%20Pages-,id%2Dtoken,-%3A%20write%20%20%20%23 (though I didn't notice the code block was truncated in my browser... sorry

Comment (junhaoliao) [reply .github/workflows/deploy-gh-pages.yml:33]: This is where the token is used: https://docs.github.com/en/rest/pages/pages?apiVersion=2022-11-28#:~:text=Default%3A%20GITHUB_SHA-,oidc_token,-string%20Required

Comment (kirkrodrigues) [reply .github/workflows/deploy-gh-pages.yml:33]: Makes sense.

**Review by kirkrodrigues — DISMISSED** [review]
For the PR title, how about:

```
ci: Add GH workflow to build app & deploy to GH pages; Update demo link to point at GH pages deployment.
```

File: .github/workflows/deploy-gh-pages.yml [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+on:
+  push:
+    branches: ["main"]
+
+jobs:
+  build:
+    runs-on: "ubuntu-latest"
+    steps:
+      - uses: "actions/checkout@v4"
```
Comment (junhaoliao) [inline .github/workflows/deploy-gh-pages.yml:11]: Shall we add
```
        with:
          submodules: "recursive"
```
here or remove it in https://github.com/y-scope/yscope-log-viewer/blob/bc8109f9f371c3327070000dee3537bdaf535428/.github/workflows/lint.yaml#L26-L27

Comment (kirkrodrigues) [reply .github/workflows/deploy-gh-pages.yml:11]: I would either add it here or leave both as is. Although we don't need it in lint.yaml currently, hopefully we'll yscope-dev-utils to this repo soon in service of adding more automated linting. So in that case, changing lint.yaml would be unnecessary; I'm not sure whether we would need to change deploy-gh-pages.yml at that point.

Speaking of, I just noticed that we should rename this workflow file with `.yaml`.

File: .github/workflows/deploy-gh-pages.yaml [hunk 10]
```diff
@@ -0,0 +1,45 @@
+name: "deploy-github-pages"
+
+on:
+  push:
+    branches: ["main"]
+
```
[the remark this answers is not in the document]
Comment (junhaoliao) [reply .github/workflows/deploy-gh-pages.yaml:6]: I pushed a fix in commit ba82b3ab71d736041ad8ba23c5819d9eb8b04a74, please review it.

File: .github/workflows/deploy-gh-pages.yaml [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    runs-on: "ubuntu-latest"
+    steps:
+      - uses: "actions/checkout@v4"
+        with:
+          submodules: "recursive"
+
+      - uses: "actions/setup-node@v4"
+        with:
+          node-version: "22"
+          cache: "npm"
```
[the remark this answers is not in the document]
Comment (junhaoliao) [reply .github/workflows/deploy-gh-pages.yaml:27]: Usage of "22" is intended.

File: .github/workflows/deploy-gh-pages.yaml [hunk 12]
```diff
@@ -0,0 +1,54 @@
+name: "deploy-github-pages"
+
+on:
+  push:
+    branches: ["main"]
+
+permissions:
+  # So the workflow can cancel in-progress jobs
+  actions: "write"
```
Comment (kirkrodrigues) [inline .github/workflows/deploy-gh-pages.yaml:9]: This is my mistake in previous workflow PRs,  but I found that this permission is not necessary for cancelling in-progress jobs. I think we should set these permissions to {} instead.

File: .github/workflows/deploy-gh-pages.yaml [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+
+concurrency:
+  group: "${{github.workflow}}-${{github.ref}}"
+  # Cancel in-progress jobs for efficiency
```
Comment (kirkrodrigues) [inline .github/workflows/deploy-gh-pages.yaml:13]: Add newline before comment.

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

File: .github/workflows/deploy-gh-pages.yaml (hunk 13 above)
Comment (junhaoliao) [inline .github/workflows/deploy-gh-pages.yaml:13]:
```suggestion

  # Cancel in-progress jobs for efficiency
```

File: .github/workflows/deploy-gh-pages.yaml (hunk 12 above)
Comment (junhaoliao) [inline .github/workflows/deploy-gh-pages.yaml:9]:
```suggestion
permissions: {}
```

**Review by kirkrodrigues — APPROVED** [review]

**Pull request merged**